### PR TITLE
Fix series and genre links containing non-ASCII characters

### LINKED ISSUE #508: viewgame links to some series or genre names containing non-ASCII characters don't work
There are two listings for "Cannelé & Nomnom" games: [beta](https://ifdb.org/viewgame?id=px4i00emsjwozza0) and [episode 1](https://ifdb.org/viewgame?id=i57zntqzl6fnkt5v).

They are marked as belonging to a series called _"Cannelé & Nomnom"_.

Clicking on the series link in either of the game pages produces (in Firefox) a _"No results were found"_ page, with the text field displaying _"series:CannelÃ© & Nomnom"_ (that's A-tilde and copyright-sign in place of the expected e-acute).

The search result link in the game page source is `/search?searchfor=series:Cannel%C3%A9+%26+Nomnom` (resolving to [this link](https://ifdb.org/search?searchfor=series:Cannel%C3%A9+%26+Nomnom) as an absolute URL). The %-encoded parts (C3 A9) represent the UTF-8 encoding of U+00E9 LATIN SMALL LETTER E WITH ACUTE. (But when interpreted as ISO-8859-1, they produce the Ã© mojibake seen above.)

If I hand-hack the URL to encode the ISO-8859-1 (latin1) representation of that character (i.e., `series:Cannel%E9+%26+Nomnom`) -- [link](https://ifdb.org/search?searchfor=series:Cannel%E9+%26+Nomnom) -- that works (returns both games from the series, sensible search box contents).

I don't know anything about IFDB's internal character representation (and I note open issue iftechfoundation/ifdb#859), but all the pages I've looked at (/viewgame, /editgame, search result) are served with HTTP header `Content-Type: text/html; charset=ISO-8859-1`, have the same in `<meta http-equiv>`, and contain literal `E9` bytes where "Cannelé" is mentioned. <s>So I guess it's generally Latin-1, and the series name in these games' data is fine, and whatever is transcoding the series name to UTF-8 for the series URL needs squashing.</s>

### PR DESCRIPTION
**Pull request by dfabulich**
Fixes #508

EDIT (Nov 7): This PR is redone from scratch now, in three commits.

The first commit trims our use of `getGameInfo` in `showuser`; we only need `$links`, and now the code shows that.

The second commit is a tiny bit scary, security-wise. `getGameInfo` used to take care of running `htmlspecialcharx` on some but not all strings it returns. Now, it doesn't pre-run `htmlspecialcharx` at all. `showuser` doesn't care, because it was only ever using the links, but now `viewgame` calls `htmlspecialchars` itself right after destructuring the result.

(I tested this by adding `<` to the series, genre, author, and system of <http://localhost:8080/viewgame?id=i57zntqzl6fnkt5v>, viewing it in XML and regular mode.)

In the second commit, we delay `htmlspecialchars` for `$system`, `$genre`, and `$seriesname`, so we can separately `urlencode` them for searching and `htmlspecialchars` encode them for display. This strategy now matches the approach we use for `$author`.

We're also now using `urlencode` and not `urlencodeFromHTML`, which was built on the assumption that the database was storing HTML entities, and that the data would be in ISO-8859-1 (Windows-1252), which it converted using [`utf8_encode`](https://www.php.net/manual/en/function.utf8-encode.php).

> utf8_encode — Converts a string from ISO-8859-1 to UTF-8
> WARNING This function has been DEPRECATED as of PHP 8.2.0. Relying on this function is highly discouraged.

### CODE REVIEW

**Comment by salty-horse** [comment]
`viewgame` handles authors [differently than the other links](https://github.com/iftechfoundation/ifdb/blob/59fe978b66c3926d9f22a2d406927b6095d472e5/www/viewgame#L1418). Should it be converted to use this function?

```php
// quote the author's name for display and use in the search HREF
$uau = urlencode($au);
$dau = htmlspecialcharx($au);
                                                                         
// show the link
echo "$conj<a href=\"search?searchfor=author%3A$uau\">$dau</a>$proLink";
```

**Comment by salty-horse** [comment]
Is #1027 (about tags with `"`) related?

**Comment by dfabulich** [comment]
I don't think so, but, to prove it, I'll fix it separately.

**Comment by dfabulich** [comment]
OK! This PR is ready for review, I think. I redid it from scratch; see the edited description at the top of the PR thread.

File: www/showuser [hunk 1]
```diff
@@ -213,6 +213,14 @@ if ($rss == 'gamenews')
              $ratingHisto, $xrefs, $inrefs) =
                  getGameInfo($db, $gameid, $curuser, false,
                              $gameErrMsg, $gameErrCode);
+        $title = htmlspecialchars($title);
```
Comment (salty-horse) [inline www/showuser:216]: `showuser` doesn't use most of these, so no need to escape them. Is this for "safety"?

Unused: `$license`, `$language`, `$system`, `$genre`, `$seriesname`, `$seriesnum`, `$forgiveness`.

(In the future I want getGameInfo to return a dict, so the caller can destructure just the fields it needs.)

File: www/viewgame [hunk 2]
```diff
@@ -1432,20 +1437,20 @@ if (!isEmpty($seriesname)) {
         echo "Part of ";
     else
         echo "Episode $seriesnum of ";
-    echo "<a href=\"search?searchfor=series:" . urlencodeFromHTML($seriesname)
-        . "\">$seriesname</a><br>";
+    echo "<a href=\"search?searchfor=series:" . urlencode($seriesname)
```
Comment (salty-horse) [inline www/viewgame:1440]: I don't remember how the search "unpacks" this data in reverse. It's not enough to `urlencode` this? It has to also be HTML-escaped?

File: www/showuser (hunk 1 above)
Comment (dfabulich) [reply www/showuser:216]: Yeah, I didn't want to bother to prove that `showuser` definitely _doesn't_ use any of these.

File: www/viewgame (hunk 2 above)
Comment (salty-horse) [reply www/viewgame:1440]: I'm not sure if you answered exactly what I was asking.

Now it's basically `urlencode(htmlspecialchars($raw_seriesname))` - is that the intention? Just to pass the data through the URL medium, it's enough to use `urlencode()` with `%XX`. Does the other side (search) expect it to also be HTML-escaped with `&XXX;`?

Comment (dfabulich) [reply www/viewgame:1440]: I claim that line 1440 is now calling `urlencode($raw_seriesname)`, and then in the body of the `<a>` tag we're displaying `htmlspecialchars($raw_seriesname)`.

The other side (search) no longer expects or requires the data to be HTML-escaped. Search just pulls out the `$seriesname` raw from the `$_REQUEST` and uses it.

File: www/showuser (hunk 1 above)
Comment (dfabulich) [reply www/showuser]: We weren't even using the `$title`! We were only ever using the `$links`. I've updated the PR to first make `showuser` _only_ grab `$links = getGameInfo(...)[20]`, which proves that we don't need to `htmlspecialchars` anything here.

File: www/viewgame (hunk 2 above)
Comment (salty-horse) [reply www/viewgame:1440]: Oh, sorry, I missed that `$seriesname` was not in the list above. OK :)

File: www/showuser [hunk 3]
```diff
@@ -197,22 +197,8 @@ if ($rss == 'gamenews')
...
-             $currentUserRating, $currentUserReview,
-             $editedbyid, $editedbyname, $moddate, $moddate2, $pagevsn,
-             $moddatelatest,
-             $historyView,
-             $dlnotes, $extReviews, $extRevDisplayRank,
-             $ratingHisto, $xrefs, $inrefs) =
-                 getGameInfo($db, $gameid, $curuser, false,
-                             $gameErrMsg, $gameErrCode);
+        $links = getGameInfo($db, $gameid, $curuser, false,
+                        $gameErrMsg, $gameErrCode)[20];
```
Comment (salty-horse) [inline www/showuser:201]: must not forget to tweak that magic 20 if more return values are added before its position. :(

**Review by salty-horse — APPROVED** [review]

File: www/showuser (hunk 3 above)
Comment (dfabulich) [reply www/showuser:201]: There are only two callers for `getGameInfo`, and messing around with argument position is obviously risky, so I'm not too worried.

**Pull request merged**